### PR TITLE
Add shard on batch mode. Als update version of torchxla2

### PR DESCRIPTION
**Pull request by qihqi**

### CODE REVIEW

File: jetstream_pt/cache_manager.py [hunk 1]
```diff
@@ -12,10 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import torch_xla2
 import jax
 import jax.numpy as jnp
 import torch
```
Comment (FanhaiLu1) [inline jetstream_pt/cache_manager.py:17]: Thanks for the change! It's more meaningful than wrap and unwrap...

File: jetstream_pt/engine.py [hunk 2]
```diff
@@ -600,23 +602,41 @@ def colocated_cpus(self) -> Union[list[engine_api.CpuDevices], None]:
...
   def get_prefix_destination_sharding(self) -> Prefix:
     """Returns the shardings necessary to transfer data between engines."""
-    return Prefix(
-        self.replicated,
-        self.cache_sharding,
-        self.replicated,
-    )
+    if self.env.shard_on_batch:
+      return Prefix(
+          self.replicated,  # cache is replicated because bs=1 for prefill
```
Comment (FanhaiLu1) [inline jetstream_pt/engine.py:607]: The multiple prefill isolated instance's performance is better than this replicated sharding (or we use samller vm to test instead of 4chips or 8 chips).

File: jetstream_pt/engine.py [hunk 3]
```diff
@@ -600,23 +602,41 @@ def colocated_cpus(self) -> Union[list[engine_api.CpuDevices], None]:
...
+          self.cache_sharding,
+          self.replicated,
+          self.replicated,
+          self.replicated,
+          self.replicated,
+          self.replicated,
+      )
+    else:
+      return DecodeState(
+          self.replicated,  # shard on batch
```
Comment (FanhaiLu1) [inline jetstream_pt/engine.py:632]: remove the comments?

File: jetstream_pt/layers.py [hunk 4]
```diff
@@ -156,10 +155,7 @@ def __call__(self, xq, xk, xv, mask, cache):
     with jax.named_scope("attn_mat1"):
       ## Attention start
       # scores = torch.einsum(jnp.einsum, "ijkl,ikml->ikjm", xq, keys) / math.sqrt(self.head_dim)
-      scores = torch_xla2.extra.call_jax(
-          jnp.einsum, "ikjl,ikml->ikjm", xq, keys
-      ) / math.sqrt(head_dim)
-      self.env.apply_sharding(scores, axis=1)
```
Comment (FanhaiLu1) [inline jetstream_pt/layers.py:162]: Is matrix transpose HLO issues been fixed?

File: jetstream_pt/engine.py [hunk 5]
```diff
@@ -86,8 +87,11 @@ def __init__(
     self.y_sharding = env.sharding_by_axis(1)
     self.x_sharding = env.sharding_by_axis(0)
     self.replicated = env.sharding_by_axis(-1)  # replicated
+
     self.cache_sharding = self.env.cache_sharding
 
+    jax.config.update("jax_enable_x64", False)
```
Comment (wang2yn84) [inline jetstream_pt/engine.py:93]: nit, we have config set up in both engine.py and the script, in the future we should place them together.

File: jetstream_pt/engine.py [hunk 6]
```diff
@@ -147,29 +151,27 @@ def _call_model_generate(
     if self.env.enable_kv_quantization:
       caches_obj = [
           cache_manager.Int8KVCacheGenerate(k, v, ks, vs, input_indexes)
-          for (k, v), (ks, vs) in torch_xla2.tensor.wrap(
-              list(zip(caches, cache_scales))
-          )
+          for (k, v), (ks, vs) in from_jax(list(zip(caches, cache_scales)))
```
Comment (wang2yn84) [inline jetstream_pt/engine.py:154]: Nit, since Pytorch should be the focus, is it better to use "to_torch" and "from_torch" instead of "from_jax" and "to_jax"? Ideally we should not see any Jax related term.

File: jetstream_pt/engine.py [hunk 7]
```diff
@@ -188,12 +190,12 @@ def _call_model_prefill(self, weights, tokens, input_indexes):
     mask = jnp.triu(mask, k=1)
     args = (tokens, input_indexes, caches, mask)
 
-    paramst, argst = torch_xla2.tensor.wrap((weights, args))
+    paramst, argst = from_jax((weights, args))
     with self._lock:
-      with torch_xla2.tensor.XLADispatchMode():
+      with torch_xla2.default_env():
```
Comment (wang2yn84) [inline jetstream_pt/engine.py:195]: For Pytorch user, DispatchMode is more descriptive than default env? And we also need some comments here why we need TorchDispatchMode here.

File: jetstream_pt/engine.py (hunk 2 above)
Comment (wang2yn84) [reply jetstream_pt/engine.py:607]: Hi Fanhai, do you mean each TPU chip do certain length of the prefill?

File: jetstream_pt/engine.py [hunk 8]
```diff
@@ -600,23 +602,41 @@ def colocated_cpus(self) -> Union[list[engine_api.CpuDevices], None]:
...
-        self.replicated,
-        self.cache_sharding,
-        self.replicated,
-        self.replicated,
-        self.replicated,
-        self.replicated,
-        self.replicated,
-    )
+    if self.env.shard_on_batch:
+      return DecodeState(
```
Comment (wang2yn84) [inline jetstream_pt/engine.py:621]: Nit, can we do 

return DecodeState(
          self.x_sharding if self.env.shard_on_batch else self.replicated,  # shard on batch
          self.cache_sharding,
          self.replicated,
          self.replicated,
          self.replicated,
          self.replicated,
          self.replicated,
      )

File: jetstream_pt/layers.py [hunk 9]
```diff
@@ -156,10 +155,7 @@ def __call__(self, xq, xk, xv, mask, cache):
...
+      scores = torch.einsum("ikjl,ikml->ikjm", xq, keys) / math.sqrt(head_dim)
```
Comment (wang2yn84) [inline jetstream_pt/layers.py:158]: Is b/329899712 fixed?

File: jetstream_pt/layers.py [hunk 10]
```diff
@@ -171,14 +167,15 @@ def __call__(self, xq, xk, xv, mask, cache):
...
+      output = torch.einsum("ikjm,ikml->ikjl", scores, values)
       if seqlen == 1:
         output = output[:, :, 0:1, :]
       # For XLA matmul performance boost
       # output = torch.matmul(scores, values)
-      self.env.apply_sharding(output, axis=1)
+      if self.env.shard_on_batch:
+        self.env.apply_sharding(output, axis=0)
+      else:
+        self.env.apply_sharding(output, axis=1)
```
Comment (wang2yn84) [inline jetstream_pt/layers.py:178]: Nit, can we use self.env.shard_on_batch to select the axis instead of duplicate the `apply_sharding`?

File: jetstream_pt/layers.py [hunk 11]
```diff
@@ -323,9 +319,14 @@ def forward(
       xk = xk.view(bsz, seqlen, self.n_kv_heads, self.head_dim)
       xv = xv.view(bsz, seqlen, self.n_kv_heads, self.head_dim)
 
-      self.env.apply_sharding(xq, axis=2)
-      self.env.apply_sharding(xk, axis=2)
-      self.env.apply_sharding(xv, axis=2)
+      if self.env.shard_on_batch:
```
Comment (wang2yn84) [inline jetstream_pt/layers.py:322]: Ditto

File: jetstream_pt/third_party/gemma/model.py [hunk 12]
```diff
@@ -148,15 +148,15 @@ def forward(
     xk = xk.view(batch_size, -1, self.num_kv_heads, self.head_dim)
     xv = xv.view(batch_size, -1, self.num_kv_heads, self.head_dim)
 
-    if self.num_kv_heads > 1:
+    if self.env.shard_on_batch:
```
Comment (wang2yn84) [inline jetstream_pt/third_party/gemma/model.py:151]: Ditto

File: run_interactive.py [hunk 13]
```diff
@@ -75,6 +75,9 @@
 _SHARDING_CONFIG = flags.DEFINE_string(
     "sharding_config", "", "config file for sharding"
 )
+_SHARD_ON_BATCH = flags.DEFINE_bool(
```
Comment (wang2yn84) [inline run_interactive.py:78]: Nit, can you add `when enabled, it overwrites the sharding_config` or something similar?

File: run_server.py [hunk 14]
```diff
@@ -89,6 +89,9 @@
 _SHARDING_CONFIG = flags.DEFINE_string(
     "sharding_config", "", "config file for sharding"
 )
+_SHARD_ON_BATCH = flags.DEFINE_bool(
```
Comment (wang2yn84) [inline run_server.py:92]: Ditto

**Review by wang2yn84 — COMMENTED** [review]
Thank you for improving the Torch xla2, adding Gemma 2 support, and clean up the code! If we can have 1 PR for each task, that'll be great and won't block you from merging.

File: jetstream_pt/layers.py (hunk 4 above)
Comment (qihqi) [reply jetstream_pt/layers.py:162]: yes

File: jetstream_pt/engine.py (hunk 2 above)
Comment (qihqi) [reply jetstream_pt/engine.py]: do you mean having 8 InterleavedEngine with one device each; vs. sharding on batch from jax?

File: jetstream_pt/engine.py (hunk 3 above)
Comment (qihqi) [reply jetstream_pt/engine.py]: done.

File: jetstream_pt/engine.py (hunk 6 above)
Comment (qihqi) [reply jetstream_pt/engine.py]: done

File: jetstream_pt/engine.py (hunk 7 above)
Comment (qihqi) [reply jetstream_pt/engine.py]: done

File: jetstream_pt/engine.py (hunk 8 above)
Comment (qihqi) [reply jetstream_pt/engine.py]: done.,

File: jetstream_pt/layers.py (hunk 9 above)
Comment (qihqi) [reply jetstream_pt/layers.py:158]: yes (hence the updating the version of torch_xla2)

File: jetstream_pt/layers.py (hunk 10 above)
Comment (qihqi) [reply jetstream_pt/layers.py]: done.

File: jetstream_pt/layers.py (hunk 11 above)
Comment (qihqi) [reply jetstream_pt/layers.py]: donee

File: jetstream_pt/third_party/gemma/model.py (hunk 12 above)
Comment (qihqi) [reply jetstream_pt/third_party/gemma/model.py]: done

File: run_interactive.py (hunk 13 above)
Comment (qihqi) [reply run_interactive.py:78]: done

File: run_server.py (hunk 14 above)
Comment (qihqi) [reply run_server.py:92]: done

File: jetstream_pt/engine.py (hunk 2 above)
Comment (FanhaiLu1) [reply jetstream_pt/engine.py]: The 8 InterleavedEngine vs sharding on batch from jax.

File: jetstream_pt/layers.py (hunk 4 above)
Comment (FanhaiLu1) [reply jetstream_pt/layers.py:162]: Can you share the PR for the fix?

File: jetstream_pt/engine.py (hunk 2 above)
Comment (wang2yn84) [reply jetstream_pt/engine.py]: Of course because there is no collective operations involved. IIUC then the best config might be enable disaggregated serving and use multiple prefill engine instances instead of sharding.

File: jetstream_pt/layers.py (hunk 4 above)
Comment (wang2yn84) [reply jetstream_pt/layers.py:162]: And also update b/329899712?

Comment (qihqi) [reply jetstream_pt/layers.py:162]: https://github.com/pytorch/xla/pull/7009

**Review by lsy323 — APPROVED** [review]

**Review by FanhaiLu1 — APPROVED** [review]

File: jetstream_pt/layers.py (hunk 4 above)
Comment (wang2yn84) [reply jetstream_pt/layers.py:162]: Thank you for the fix! Not directly related, but I see you specially treated einsum to fix the issue. Please also fix b/329899713 for matmul. But the conversation in this PR can be closed now.

**Review by wang2yn84 — APPROVED** [review]
